include/swift/SIL/MemoryLocations.h

```diff
@@ -241,8 +281,18 @@ class MemoryLocations {
   /// Sets the location bits of \p addr in \p bits, if \p addr is associated
   /// with a location.
   void setBits(Bits &bits, SILValue addr) const {
-    if (auto *loc = getLocation(addr))
+    if (auto *loc = getLocation(addr)) {
       bits |= loc->subLocations;
+      if (loc->parentHasValueDeinit) {
```

Comment:
I'm wondering why this is needed.
Isn't the check to catch wrong member-wise destroying the same as for resilient structs?
Resilient struct locations just have set their `numFieldsNotCoveredBySubfields` to `INT_MAX` which always requires that the whole struct is destroyed.

Reply:
This check allows structs with deinits to be memberwise initialized. That's what the comment on `hasValueDeinit` is talking about. Ideally, we would fix DI to insert an explicit `mark_initialized` whenever value with deinitializers are fully initialized. Then we can remove the [parent]HasValueDeinit location flags. But in the meantime, we "fake" full initialization because we urgently need a way to catch deinitialization bugs.

Reply:
I see. But this could be solved by setting the "self" bit directly at the `alloc_stack`. For example:

```
  %0 = alloc_stack $MoveOnly                   // bits = 01  (bit 0 = self, bit 1 = singleField)
  %1 = struct_element_addr %0, #singleField
  store %initVal to %1                         // bits = 11  => okay, fully initialized
```

and

```
  %1 = struct_element_addr %0, #singleField    // bits = 11
  destroy_value %1                             // bits = 01
  dealloc_stack %0                             // error! self not destroyed
```

Reply:
The `self` bit is set by alloc stack when a deinit is present:
https://github.com/apple/swift/pull/66447/files#diff-80beed39aa95523c6885509900bd19d20cc859ddfb34a0469a83a4131fc52b98R420
That's needed just to handle empty structs. I think that actually needs to be expanded to handle `@out` as well.

The problem is that structs that are not empty can be reinitialized after they are destroyed. The only indication that they are reinitialized is the store to a field.

Yes, this is bad SIL design. But for now we just need a way to resolve incorrect deinitialization bugs and stop arguing over which pass is doing the wrong thign.

Reply:
Still, I don't understand why you can't express this constraints with simple bit values. Why is this special case (`if (loc->parentHasValueDeinit)`) in `setBits` needed?

Reply:
Without the special case, the verifier would fail to recognize reinitialization of the aggregate's 'self' bit.

The special case is cached with `parentHasValueDeinit`. Otherwise every time a bit is set, the verifier would need to check all parents to see if a deinit is present. This special case happens extremely rarely, so there's not reason to complicate the normal verification path.

```
struct PairDeinit : ~Copyable {
  var a: T
  var b: T

  deinit
}

// The aggregate is implicitly reinitialized by setting its fields.
// The verifier does not yet ensure that all fields are reinitialized;
// instead, we should add a mark_initialized instruction to SIL
// to handle reinitialization of values with deinits.
sil [ossa] @testPairInit : $(@owned T) -> () {
bb0(%0 : @owned $T):
  %addr = alloc_stack $PairDeinit

  %2 = copy_value %0 : $T
  %3 = struct_element_addr %addr : $*PairDeinit, #PairDeinit.a
  store %2 to [init] %3 : $*T
  %5 = copy_value %0 : $T
  %6 = struct_element_addr %addr : $*PairDeinit, #PairDeinit.b
  store %5 to [init] %6 : $*T
  destroy_addr %addr : $*PairDeinit

  %9 = copy_value %0 : $T
  %10 = struct_element_addr %addr : $*PairDeinit, #PairDeinit.a
  store %9 to [init] %10 : $*T
  %12 = struct_element_addr %addr : $*PairDeinit, #PairDeinit.b
  store %0 to [init] %12 : $*T
  destroy_addr %addr : $*PairDeinit

  dealloc_stack %addr : $*PairDeinit
  %retval = tuple ()
  return %retval : $()
}
```

Reply:
okay

Reply:
We've decided not to allow memberwise reinitialization for now. So the only reason that we would need the parentHasDeinit flag is if some type of initializer requires it. If initializers always have an alloc_stack, then we can get rid of the flag.